mlflow/models/evaluation/base.py

```diff
@@ -341,6 +341,7 @@ def __init__(self, data, *, targets, name=None, path=None, feature_names=None):
                     f"feature_{str(i + 1).zfill(math.ceil((math.log10(num_features + 1))))}"
                     for i in range(num_features)
                 ]
+            self._original_column_names = list(range(num_features))
```

Comment:
Do we need this change? We no longer truncate feature names (see https://github.com/mlflow/mlflow/pull/6721). I don't think we need to memorize `_original_column_names`.

@WeichenXu123 Thoughts?

Reply:
If the test data is a NumPy array, this fix will prevent the warnings from sklearn

Reply:
We can't just use self.feature_names as they are strings, which introduces the named columns producing the sklearn warnings when using explainability

Reply:
@WeichenXu123 Can we directly pass a numpy array to shap explainer?

Reply:
@WeichenXu123

Reply:
> Can we pass a numpy array to shap explainer?

I think no. Because some Sklearn model require inference data with column names (e.g. some transformers in a pipeline need to read column name)

Reply:
if original data is a dataframe with unlabeled features,
when calling shap evalaution, we should restore the original dataframe when call model.predict.

Reply:
Can we just pass the data supplied by the user without auto-generating feature names?

Reply:
This case `pd.DataFrame(np_array)` it generates unlabeled dataframe. This is valid case.
Or we can raise error when constructing EvaluationDataset if input dataset is unlabeled Dataframe

Reply:
When do users do `pd.DataFrame(numpy.array)`?

Reply:
It comes from @dsgibbons 's case.
If it is a rare case, we can forbid this case when constructing EvaluationDataset

Reply:
I don't think we need to forbid. We can just pass the dataframe to shap without altering the columns.

Reply:
I remember trying this @harupy but I believe this led to issues with the categorical data test. I can have another look.

Reply:
Yep, I can confirm that if you just try to pass `pd.DataFrame` directly to shap without adding in column information, the test `test_pipeline_model_kernel_explainer_on_categorical_features` fails due to:

```*** KeyError: "None of [Index(['f1', 'f2'], dtype='object')] are in the [columns]"``` 

I'm not sure why, but it seems as though this test needs explicit column naming when using shap. 

The only other obvious fix (other than this PR) I can see for this problem:

1. Unstringify the feature names, i.e., replace
```
self._feature_names = [
    generate_feature_name_if_not_string(c) for c in self._features_data.columns
]
```
with 
```
self._feature_names = self._features_data.columns
```

2. Revert to `master`'s use of `feature_names` instead of `original_column_names` e.g.,
```
def _shap_predict_fn(x, predict_fn, feature_names):
    return predict_fn(_get_dataframe_with_renamed_columns(x, feature_names))
```

3. Stringify feature names after evaluation and just before the shap plots get created.
```
shap_values.feature_names = [generate_feature_name_if_not_string(n) for n in self.feature_names]
```

I can make another commit that implements these changes. If you don't like any of these changes and think that fixing this edge case warning makes the code too complicated then I can revert all changes relating to the sklearn warnings and just leave the naming fix for `explainability_ignore_exceptions`

Reply:
To simplify code and avoid sklearn warning / errors, what about:

When constructing EvaluationDataset, if it is pandas dataframe input, disallow unlabeled dataframe input.

When doing shap evaluation, keep calling `model.predict` with original input data (original numpy array or original pandas dataframe)

Reply:
```
the test test_pipeline_model_kernel_explainer_on_categorical_features fails due to:

*** KeyError: "None of [Index(['f1', 'f2'], dtype='object')] are in the [columns]"
```

This is because the pipeline trained with labeled dataframe, and it contains some transformers which need to read the column name, so in prediction it also need column name.

Reply:
I'm not sure how to call `model.predict` with original input data in a way that still passes  `test_pipeline_model_kernel_explainer_on_categorical_features`? Do you have any suggestions?

Reply:
I personally preferred the previous solution of saving the original column names, as that also silenced the sklearn warning when evaluating on np.ndarrays (not just unlabeled dataframes). For example,  `test_regressor_evaluation[linear_regressor_model_uri]`